Don't go through init sequence twice

kohsuke opened the pull request:
As a part of #1847, 3b5424e507be introduced a second invocation to
executeReactor. This is causing a problem in that it exectes
InitMilestone progression twice. If you are looking at
Jenkins.getInitLevel() it will go through STARTED ... COMPLETED once,
then it goes back to STARTED and then go to COMPLETED again.

And this code looks bit odd anyway. Why execute a whole reactor
just to make one method invocation? But I didn't see any discussion of
this in PR 1847.

This change fixes this oddity by making this a part of the main
(and sole) initialization reactor session. It's not clear what
milestone should be the pre-requisite, but given that it's only looking
at dependency chain, PLUGINS_PREPARED should be adequate.

@reviewbybees in particular @tfennelly who wrote the original change

kohsuke committed:
Don't go through init sequence twice

As a part of PR 1847, 3b5424e507be introduced a second invocation to
executeReactor. This is causing a problem in that it exectes
InitMilestone progression twice. If you are looking at
Jenkins.getInitLevel() it will go through STARTED ... COMPLETED once,
then it goes back to STARTED and then go to COMPLETED again.

And this code looks bit odd anyway. Why execute a whole reactor
just to make one method invocation? But I didn't see any discussion of
this in PR 1847.

This change fixes this oddity by making this a part of the main
(and sole) initialization reactor session. It's not clear what
milestone should be the pre-requisite, but given that it's only looking
at dependency chain, PLUGINS_PREPARED should be adequate.

reviewbybees commented:
This pull request originates from a [CloudBees](https://www.cloudbees.com/) employee. At CloudBees, we require that all pull requests be reviewed by other CloudBees employees before we seek to have the change accepted. If you want to learn more about our process please see [this explanation](https://github.com/reviewbybees/about#about-reviewbybees).

kzantow commented:
@kohsuke there is a related test failure on this PR, other than that :+1:

oleg-nenashev commented:
Yes, it seems the test failure is really caused by this change. Needs to be solved

kohsuke commented:
Huh, interesting. Not obvious to me how this failure relates to this change, but let me see.

kohsuke committed:
Fixed a test regression

Previously, wa.contextDestroyed(null) interrupts the Jenkins.<init>
thread, and it was the 2nd reactor call that was throwing InterruptedException
from this call. Now that I removed that 2nd reactor call, the init
thread never actually fails.

The intent of the test case is to make sure the boot failure hook is
invoked, so even with this change it covers the original intent.

tfennelly commented:
@kohsuke I'm fine with these changes if you are :bee: . I obviously didn't understand the whole reactor architecture in the first place, so ... :smiley:

tfennelly commented:
Closing and reopening to re-trigger build.

kzantow commented:
:bee: only the normal erroneous test failures, now

jglick commented on test/src/test/java/hudson/util/BootFailureTest.java:
```diff
@@ -154,6 +154,8 @@ public File allocate() throws Exception {
         public void onLoaded() {
             wa.contextDestroyed(null);
+            // make the Jenkins.<init> thread abort
+            throw new Error();
```
Not going to work once we fix `ItemListener` to catch errors from `onLoaded` implementations, though that is also true of `InduceBootFailure`.

But the :bug: is because the fact you needed to do this suggests to me that you somehow regressed the logic in `WebAppMain.contextDestroyed` which is supposed to be interrupting the `initThread` and which this test (from #1950) was verifying.

kohsuke replied:
I don't think so. See https://github.com/jenkinsci/jenkins/pull/2177/commits/1f051eb37e3c33da2ed7a1d51f08ef572c4e085e where I discuss this change. `WebAppMain.contextDestroyed` does interrupt the init thread, but the problem is that the init thread isn't waiting by this point in `Jenkins.<init>` that invokes this callback. So the interrupt will never get processed.

This test case from get-go incorrectly relied on the presence of the `Object.wait()` code in the init thread after this callback, which is no longer true.

kohsuke committed:
Treat any interruption during boot as a fatal failure
